R/convert_output_types.R

```diff
@@ -55,31 +55,59 @@ convert_output_type <- function(model_out_tbl, terminal_output_type_id) {
-  task_id_cols <- subset_task_id_names(names(model_out_tbl))
+  model_out_cols <- colnames(model_out_tbl)
+  task_id_cols <- subset_task_id_names(model_out_cols)
   grouped_model_outputs <- dplyr::group_by(model_out_tbl, dplyr::across(dplyr::all_of(c("model_id", task_id_cols))))
   terminal_output_type_id <- terminal_output_type_id[[terminal_output_type]]
 
-  if (terminal_output_type == "mean") {
-    converted_outputs_tmp <- dplyr::reframe(
-      grouped_model_outputs,
-      output_type_id = terminal_output_type_id,
-      value = mean(.data[["value"]])
-    )
-  } else if (terminal_output_type == "median") {
-    converted_outputs_tmp <- dplyr::reframe(
-      grouped_model_outputs,
-      output_type_id = terminal_output_type_id,
-      value = stats::median(.data[["value"]])
-    )
+  if (terminal_output_type %in% c("mean", "median")) {
+    otid_cols <- "output_type_id"
+    transform_fun <- get(terminal_output_type)
+    transform_args <- list(x = quote(.data[["value"]]))
+    grouped_model_outputs <- grouped_model_outputs
   } else if (terminal_output_type == "quantile") {
-    converted_outputs_tmp <- dplyr::reframe(
+    otid_cols <- "output_type_id"
+    transform_fun <- stats::quantile
+    transform_args <- list(
+      x = quote(unlist(.data[["initial_value"]])),
+      probs = quote(unlist(.data[["output_type_id"]])),
+      names = FALSE
+    )
+    grouped_model_outputs <- dplyr::summarize(
       grouped_model_outputs,
-      output_type_id = terminal_output_type_id,
-      value = stats::quantile(.data[["value"]], as.numeric(terminal_output_type_id), names = FALSE)
+      initial_value = list(.data[["value"]])
     )
   }
 
-  dplyr::mutate(converted_outputs_tmp, output_type = terminal_output_type, .before = "output_type_id")
+  # if overlapping columns, join grouped_model_outputs with terminal_output_type_id
+  # else, we mutate a new nested column, then unnest in that case
+  join_cols <- task_id_cols[task_id_cols %in% colnames(terminal_output_type_id)]
+  if (length(join_cols) > 0) {
+    grouped_model_outputs <- grouped_model_outputs |>
+      dplyr::left_join(
+        terminal_output_type_id,
+        by = join_cols,
+        relationship = "many-to-many"
+      )
+  } else {
+    grouped_model_outputs <- grouped_model_outputs |>
+      dplyr::mutate(output_type_id = list(terminal_output_type_id)) |>
+      tidyr::unnest(cols = "output_type_id")
+  }
+
+  # compute prediction values, clean up included columns
+  grouped_model_outputs |>
+    dplyr::group_by(dplyr::across(dplyr::all_of(c("model_id", task_id_cols, otid_cols)))) |>
+    dplyr::summarize(
+      value = list(do.call(transform_fun, args = transform_args)),
+      .groups = "drop"
+    ) |>
+    tidyr::unnest(cols = "value") |>
+    dplyr::mutate(
+      output_type = terminal_output_type,
+      .before = "output_type_id"
+    ) |>
+    dplyr::select(dplyr::all_of(model_out_cols))
```

Comment:
Again, if you were to use the non-nesting approach, this would be the strategy: drop the output_type and output_type_id columns and let the groups do all of the work. 

```suggestion
  if (length(join_cols) > 0) {
    grouped_model_outputs <- grouped_model_outputs |>
      dplyr::select(-.data[["output_type"]], -.data[["output_type_id"]]) |>
      dplyr::left_join(
        terminal_output_type_id,
        by = join_cols,
        relationship = "many-to-many"
      )
  }

  # compute prediction values, clean up included columns
  grouped_model_outputs |>
    dplyr::group_by(dplyr::across(dplyr::all_of(c("model_id", task_id_cols, otid_cols)))) |>
    dplyr::summarize(
      value = do.call(transform_fun, args = transform_args),
      .groups = "drop"
    ) |>
    dplyr::mutate(
      output_type = terminal_output_type,
      .before = "output_type_id"
    ) |>
    dplyr::select(dplyr::all_of(model_out_cols))
```